File: src/routes/judge.js
const express = require('express');
const createError = require('http-errors');
const Score = require('../models/score');
const Lab = require('../models/lab');
const User = require('../models/user');
const judgeapi = require('../utils/judgeapi');
const path = require('path');
const fs = require('fs');
const os = require('os');
const auth = require('../middlewares/auth');
const multer  = require('multer')
const router = express.Router();

//const uploadPath = path.join(__dirname, '../files');
const upload = multer({ dest: os.tmpdir() });
//const maxUploadCount = 20;

router.post('/', auth.checkSignIn, upload.any(), async function(req, res, next) {
  try {
    const username = req.session.user.username;
    const userdata = await User.getUser({username});
    if(!userdata) throw createError(404);
    const lab = await Lab.getLab(req.body.id);
    if(!lab) throw createError(404);
    let allow = lab.promissions.includes("all");
    userdata.groups.forEach(function(group) {
      if(lab.promissions.includes(group)) {
          allow = true;
      }
    });
    if(!allow) throw createError(404);
    const ipindex = userdata.ipindex;

    // send request to judge server
    const body = {
      labId: lab.id,
      username,
      ipindex,
      data: await placeDatas({"input":(input) => {
        return new Promise((resolve, reject) => {
          resolve(input || "");
        });
      }, "upload":(file) => {
        if (!file) throw createError(400, 'The number of files did not match');
        return new Promise((resolve, reject) => {
          fs.readFile(path.join(os.tmpdir(), file.filename), (err, data) => {
            if (err) {
              reject(err);
            }
            resolve(Buffer.from(data).toString('base64'));
          });
        });
      }}, lab),
    };
    const result = await judgeapi.post("score/judge", body);
    if(!result.alive) throw createError(404, 'Api server is not alive');
    if(!result.data.alive || !result.data.results) 
    {
      res.send({ alive: false });
      return;
    }
    let score = calcScore(result.data.results);

    // save score
    await Score.addScore(username, lab.id, score, result.data.results);

    for(var i = 0; i < lab.deadlines.length; i++) {
      if(Date.now() < Date.parse(lab.deadlines[i].time)) {
        sorce = score * lab.deadlines[i].score;
        break;
      }
    }

    res.send({ alive: result.data.alive, score, results: result.data.results, stdout: result.data.stdout, stderr: result.data.stderr });
  } catch(err) {
    next(err);
  }
  removeTempFiles(req.files);
});

router.get('/canjudge', auth.checkSignIn, async function(req, res, next) {
  try {
    const username = req.session.user.username;
    const result = await judgeapi.post("score/canjudge", {username});
    if(!result.alive) throw createError(404, 'Api server is not alive');
    else res.send(result.data);
  } catch(err) {
    next(err);
  }
});

function calcScore(judgeResult) {
  let score = 0;
  for (var key in judgeResult) {
    judgeResult[key].forEach((result) => {
      if (result.ans === true) {
        score += parseInt(result.weight);
      }
    })
  }
  return score;
}

async function placeDatas(solve, lab) {
  return Promise.all(lab.contents.map(async (content) => {
    return {
      type: content.type,
      name: content.name,
      data: await solve[content.type](data[content.type][content.name]),
    };
  }));
}

function removeTempFiles(files) {
  files.forEach((file) => {
    fs.rm(file.path, (err) => {
      if (err) {
        console.error(err);
      }
    });
  });
}

module.exports = router;

File: router/judge/judge.go
package judge
import (
    "io/ioutil"
    "encoding/base64"

    "github.com/gin-gonic/gin"

    "github.com/NCKU-NASA/nasa-judge-lib/schema/user"
    "github.com/NCKU-NASA/nasa-judge-lib/schema/lab"
    "github.com/NCKU-NASA/nasa-judge-lib/schema/lab/content"
    contenttype "github.com/NCKU-NASA/nasa-judge-lib/enum/content_type"

    "github.com/NCKU-NASA/nasa-judge-backend/apis/judge"
    "github.com/NCKU-NASA/nasa-judge-backend/middlewares/auth"
    "github.com/NCKU-NASA/nasa-judge-backend/utils/errutil"
)

var router *gin.RouterGroup

func Init(r *gin.RouterGroup) {
    router = r
    router.POST("/", auth.CheckSignIn, onjudge)
    router.GET("/canjudge", auth.CheckSignIn, canjudge)
}

func canjudge(c *gin.Context) {
    userdata, _ := c.Get("user")
    judge.CanJudge(c, userdata.(user.User))
}

func onjudge(c *gin.Context) {
    form, _ := c.MultipartForm()
    userdata, _ := c.Get("user")
    labId := c.PostForm("id")
    labdata, err := lab.GetLab(labId)
    if err != nil {
        errutil.AbortAndStatus(c, 400)
        return
    }
    var contents content.Contents
    for key, value := range c.Request.PostForm {
        if key == "id" {
            continue
        }
        contents = append(contents, content.Content{
            Type: contenttype.Input,
            Name: key,
            Data: value[0],
        })
    }
    for key, value := range form.File {
        f, err := value[0].Open()
        if err != nil {
            errutil.AbortAndStatus(c, 500)
            return
        }
        result, err := ioutil.ReadAll(f)
        if err != nil {
            errutil.AbortAndStatus(c, 500)
            return
        }
        contents = append(contents, content.Content{
            Type: contenttype.Upload,
            Name: key,
            Data: base64.StdEncoding.EncodeToString(result),
        })
    }
    nowscore, err := judge.Judge(labdata, userdata.(user.User), contents)
    if err != nil {
        errutil.AbortAndStatus(c, 500)
        return
    }
    nowscore.User = nil
    nowscore.Lab = nil
    nowscore.Data = nil
    c.JSON(200, nowscore)
}
